v2/api/keyvault/customizations/vault_extensions.go

```diff
@@ -0,0 +1,376 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+package customizations
+
+import (
+	"context"
+	"strings"
+
+	keyvault "github.com/Azure/azure-service-operator/v2/api/keyvault/v1api20210401previewstorage"
+	resources "github.com/Azure/azure-service-operator/v2/api/resources/v1api20200601storage"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/arm"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/runtime"
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"sigs.k8s.io/controller-runtime/pkg/conversion"
+
+	"github.com/Azure/azure-service-operator/v2/internal/genericarmclient"
+	"github.com/Azure/azure-service-operator/v2/internal/reflecthelpers"
+	"github.com/Azure/azure-service-operator/v2/internal/resolver"
+	"github.com/Azure/azure-service-operator/v2/internal/util/kubeclient"
+	"github.com/Azure/azure-service-operator/v2/internal/util/to"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime/extensions"
+)
+
+var _ extensions.ARMResourceModifier = &VaultExtension{}
+
+const (
+	CreateMode_Default         = "default"
+	CreateMode_Recover         = "recover"
+	CreateMode_CreateOrRecover = "createOrRecover"
+	CreateMode_PurgeThenCreate = "purgeThenCreate"
+)
+
+// ModifyARMResource implements extensions.ARMResourceModifier.
+func (ex *VaultExtension) ModifyARMResource(
+	ctx context.Context,
+	armClient *genericarmclient.GenericClient,
+	armObj genruntime.ARMResource,
+	obj genruntime.ARMMetaObject,
+	kubeClient kubeclient.Client,
+	resolver *resolver.Resolver,
+	log logr.Logger,
+) (genruntime.ARMResource, error) {
+
+	kv, ok := obj.(*keyvault.Vault)
+	if !ok {
+		return nil, errors.Errorf(
+			"Cannot run VaultExtension.ModifyARMResource() with unexpected resource type %T",
+			obj)
+	}
+
+	// Type assert that we are the hub type. This will fail to compile if
+	// the hub type has been changed but this extension has not been updated to match
+	var _ conversion.Hub = kv
+
+	// If createMode is nil, nothing for us to do
+	if kv.Spec.Properties == nil {
```

Comment:
Do we also need to add check for `CreateMode` here? Something like `kv.Spec.Properties == nil || kv.Spec.Properties.CreateMode == nil` ?